clients/drcachesim/tests/record_filter_unit_tests.cpp

```diff
@@ -489,6 +510,250 @@ test_chunk_update()
     return true;
 }
 
+static bool
+test_trim_filter()
+{
+    constexpr addr_t TID = 5;
+    constexpr addr_t PC_A = 0x1234;
+    constexpr addr_t ENCODING_A = 0x4321;
+    constexpr addr_t PC_B = 0x5678;
+    constexpr addr_t ENCODING_B = 0x8765;
+    {
+        // Test invalid parameters.
+        auto filter = std::unique_ptr<record_filter_func_t>(
+            new dynamorio::drmemtrace::trim_filter_t(150, 149));
+        if (filter->get_error_string().empty()) {
+            fprintf(stderr, "Failed to return an error on invalid params");
+            return false;
+        }
+        auto filter2 = std::unique_ptr<record_filter_func_t>(
+            new dynamorio::drmemtrace::trim_filter_t(150, 150));
+        if (filter2->get_error_string().empty()) {
+            fprintf(stderr, "Failed to return an error on invalid params");
+            return false;
+        }
+    }
+    {
+        // Test removing from mid-way in the 1st chunk to the very end.
+        std::vector<test_case_t> entries = {
+            // Header.
+            { { TRACE_TYPE_HEADER, 0, { 0x1 } }, true, { true } },
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_VERSION, { 0x2 } }, true, { true } },
+            { { TRACE_TYPE_MARKER,
+                TRACE_MARKER_TYPE_FILETYPE,
+                { OFFLINE_FILE_TYPE_ENCODINGS } },
+              true,
+              { true } },
+            { { TRACE_TYPE_THREAD, 0, { TID } }, true, { true } },
+            { { TRACE_TYPE_PID, 0, { 0x5 } }, true, { true } },
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CACHE_LINE_SIZE, { 0x6 } },
+              true,
+              { true } },
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CHUNK_INSTR_COUNT, { 2 } },
+              true,
+              { true } },
+            // Chunk 1.
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 100 } },
+              true,
+              { true } },
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_CPU_ID, { 0 } }, true, { true } },
+            { { TRACE_TYPE_ENCODING, 2, { ENCODING_A } }, true, { true } },
+            { { TRACE_TYPE_INSTR, 2, { PC_A } }, true, { true } },
+            // Removal starts here.
+            { { TRACE_TYPE_MARKER, TRACE_MARKER_TYPE_TIMESTAMP, { 200 } },
+              true,
+              { false } },
```

Comment:
Keep all the timestamps in the removed region?  Or just the single boundary timestamp?  The trim-before does keep the boundary timestamp there; so you mean a change to keep just the endpoint timestamp before the end-trimming starts?

Reply:
> so you mean a change to keep just the endpoint timestamp before the end-trimming starts?

Yes.

Reply:
OK, this will require a little state in the trimmer to delay the remove-state change.  I do plan to do this but just ran out of time today and with a bunch of other review requests and another round anyway I'm uploading what I have at this point.

Reply:
Done, and updated the tests.